First draft of 3.1.1 blogpost (#1324)
scala/scala-lang: blog/_posts/2022-01-17-scala-3.1.1-released.md

---
layout: blog-detail
post-type: blog
by: Paweł Marks, VirtusLab
title: Scala 3.1.1 and forward compatibility news
---

Hello from the Scala 3 team! After a slightly longer than usual pause, we are back with a new release of the Scala compiler and exciting news about improvements in forward compatibility. The delay in the release was caused by urgent bugs that needed backporting and testing, and then the holiday season.

## Improvements is Scala 3.1.1

Scala 3.1.1-RC2 was promoted to a full stable release of the language. Scala 3.1.1 is the first patch version after the 3.1 minor release. It is therefore focused on bug fixes and improvements in overall correctness and stability. Although there are no new additions to the language, you can still spot some noticeable improvements:

> - You can use `using` clauses to introduce Scala 2 conversions (`<:<` operator).

Reviewer: Maybe add links to PRs?

Reviewer 2: Agree this needs further explanation, perhaps just in the form of a link; I cannot make sense of the description

> - Mirrors can be correctly summoned for sum types compiled by Scala 3.0.x.

Reviewer 3: ```suggestion
- Mirrors can be correctly summoned for hierarchical sum types compiled by Scala 3.0.x.
```

- Scaladoc documentation has a new, improved UI.

For a comprehensive list of fixes, take a look at the full changelog on GitHub. // TODO: link

## 3.1.2-RC1 and forward compatibility improvements

Scala 3 has excellent backward compatibility guarantees between the minor versions. On the other hand, after the recent release of Scala 3.1, we can see that the libraries should be really cautious with updating the compiler version, as it is forcing the bump on every user of that library. We do not want library authors to be stuck on old versions of the compiler as that would mean that they are locked out of many bugfixes, or we would need to spend enormous effort on backporting every bugfix to all past versioning lines.

We are sure we want to mitigate those difficulties. We already have implemented the first and most important part of the solution and have it ready for public testing. Beginning in Scala 3.1.2-RC1, that was just released, compiler is able to generate output that can be consumed by older versions. To enable that, you need to specify a minimum version of the language using the experimental `-Yscala-release` flag.

> So, for example, compiling your library using Scala 3.1.2 with `-Yscala-release=3.0` flag, you make it possible for people still using Scala 3.0 to depend on your library. This has one limitation, you won't be able to use any symbol added to the standard library after 3.0. They are marked in the code and in the documentation by the internal `@Since` annotation that was also added in this version.

Reviewer: I think we should use `3.1.2-RC1` here instead of just `3.1.2` for consistency.

> So, for example, compiling your library using Scala 3.1.2 with `-Yscala-release=3.0` flag, you make it possible for people still using Scala 3.0 to depend on your library. This has one limitation, you won't be able to use any symbol added to the standard library after 3.0. They are marked in the code and in the documentation by the internal `@Since` annotation that was also added in this version.

Reviewer: The syntax for passing the option from the command line is `-Yscala-release 3.0` or `-Yscala-release:3.0`.  This won't work with `=` as far as I know.

> So, for example, compiling your library using Scala 3.1.2 with `-Yscala-release=3.0` flag, you make it possible for people still using Scala 3.0 to depend on your library. This has one limitation, you won't be able to use any symbol added to the standard library after 3.0. They are marked in the code and in the documentation by the internal `@Since` annotation that was also added in this version.

Reviewer: It's lower case `@since`

> Those restrictions are not final, and in the future, we may decide to allow using some functions from the newer versions of the standard library and to make the `@Since` annotation public.

Reviewer: `@since`

---
layout: blog-detail
post-type: blog
by: Paweł Marks, VirtusLab
title: Scala 3.1.1 and forward compatibility news
---

> Hello from the Scala 3 team! After a slightly longer than usual pause, we are back with a new release of the Scala compiler and exciting news about improvements in forward compatibility. The delay in the release was caused by urgent bugs that needed backporting and testing, and then the holiday season.

Reviewer 2: I think you could omit the material about timing. It's normal for releases to slip a bit, especially around the holidays. People are primarily interested in the content of the release; few care whether it's a bit late.

If you feel you must account for the lateness, consider doing so much nearer end of the announcement, rather than the beginning. Leading with this sets a defensive, things-aren't-going-well tone. But things actually are going well (the delay is unimportant), so there's no need to set that tone.

## Improvements is Scala 3.1.1

> Scala 3.1.1-RC2 was promoted to a full stable release of the language. Scala 3.1.1 is the first patch version after the 3.1 minor release. It is therefore focused on bug fixes and improvements in overall correctness and stability. Although there are no new additions to the language, you can still spot some noticeable improvements:

Reviewer 2: Any mention of Scala 3.1.1-RC2 is supplementary material to omit or move to later, not important information to lead with.

"Scala 3.1.1 is the first patch version after the 3.1 minor release." doesn't add any meaning as far as I can see, I think you could just omit it. Or maybe condense to "Scala 3.1.1 is a patch release focused on bugfixes." I don't think "improvements in overall correctness and stability" adds any meaning and can be omitted.

> - You can use `using` clauses to introduce Scala 2 conversions.

Reviewer 2: This is confusing — I genuinely don't know what this means.

When you say "conversions", you mean "implicit conversions", or you mean something else?

Is `<:<` a "Scala 2 conversion"? (If so, why? Scala 3 has `<:<` too.)

Author: Yeah, I wanted to write, that now under the hood `<:<` creates implicit conversion inside the body of the function. Also, I wanted to write that now it works the same way as in scala 2, and I'm not sure what else was I thinking, since I finished with writing neither of that. I hope it makes sense now.

  Now the code like this should compile and perform necessary conversion from `A` to `B` in the method body:

  ```scala
    def example[A, B](a: A)(using A <:< B) =
      val b: B = a
      ???
  ```

  To know more see [PR #13662](https://github.com/lampepfl/dotty/pull/13662) and [related issue](https://github.com/lampepfl/dotty/issues/12955).

> - Mirrors can be correctly summoned for hierarchical sum types compiled by Scala 3.0.x ([PR #14100](https://github.com/lampepfl/dotty/pull/14100)).

Reviewer 2: If this is important enough to release note, I think the wording could do more to convey what the actual importance of this is.

Author: @bishabosha Does that make sense now?

- Dual Scala 2/3 macros can be now declared in non-experimental scopes ([PR #13795](https://github.com/lampepfl/dotty/pull/13795)).

> - Scaladoc can correctly handle symbolic class names, such as `/` or `\` ([PR #13827](https://github.com/lampepfl/dotty/pull/13827)).

Reviewer 2: too minor to merit release-noting, IMO

> - Scaladoc documentation has a new, improved UI.

Reviewer 2: is it "new", or is it merely "improved"?

is there a link with more information?

Author: The only link is a pr without any before and after comparison so ¯\_(ツ)_/¯.

Reviewer 2: heh. I'd go with "improved" then, or perhaps just omit the bullet entirely, if you aren't confident the changes are truly notable

For a comprehensive list of fixes, take a look at the full changelog on GitHub. // TODO: link

## 3.1.2-RC1 and forward compatibility improvements

> Scala 3 has excellent backward compatibility guarantees between the minor versions. On the other hand, after the recent release of Scala 3.1, we can see that the libraries should be really cautious with updating the compiler version, as it is forcing the bump on every user of that library. We do not want library authors to be stuck on old versions of the compiler as that would mean that they are locked out of many bugfixes, or we would need to spend enormous effort on backporting every bugfix to all past versioning lines.

Reviewer 2: Link to somewhere I can read about this?

Author: @prolativ Are you going to publish your documentation of `-Yscala-release`?

Reviewer: Waiting for someone's review
https://github.com/scala/docs.scala-lang/pull/2301/files

> We are sure we want to mitigate those difficulties. We already have implemented the first and most important part of the solution and have it ready for public testing. Beginning in Scala 3.1.2-RC1, that was just released, compiler is able to generate output that can be consumed by older versions. To enable that, you need to specify a minimum version of the language using the experimental `-Yscala-release` flag.

Reviewer 2: This reads strangely. Oh, there's a Scala 3.1.2-RC1 now? But I'm only finding that it in the middle of a paragraph, in the middle of a sentence, in an offhand phrase like "that was just released"?

I'd suggest leading with announcing the availability of 3.1.2-RC1, then go into the material about compatibility.

So, for example, compiling your library using Scala 3.1.2-RC2 with `-Yscala-release:3.0` flag, you make it possible for people still using Scala 3.0 to depend on your library. This has one limitation, you won't be able to use any symbol added to the standard library after 3.0. They are marked in the code and in the documentation by the internal `@since` annotation that was also added in this version.

Those restrictions are not final, and in the future, we may decide to allow using some functions from the newer versions of the standard library and to make the `@since` annotation public.

> One important note is that the `-Yscala-release` flag is experimental and not yet available in any stable release of the compiler. It means the build tools, such as SBT or Mill, are not yet aware of its existence. Of course, you can declare usage of the flag in your build file, and the compiler will respect it. However, there may still be some hiccups, like generated artifacts depending on the incorrect version of the stdlib. We will do our best to help fix those minor issues before the stable release of 3.1.1.

Reviewer 2: So what's the upshot for library authors? I'm a library author; what do I do, what do you recommend, what is expected of me, what is the upshot, what do I actually need to know about this right now?

I'd suggest leading with answers to those questions. _Then_ follow it up with technical details and justifications if you feel they are necessary.

Reading this, the situation seems to be rather complex and is still changing rapidly. In that case, it might actually be better to omit the technical details from the release announcement; consider giving people the TL;DR on where things stand, with a link to a contributors thread on Discourse where people can read about and discuss details.

(I might be wrong about what the right way to handle this is, but if I am wrong, that speaks to a lack of clarity in the current text, which should be addressed regardless.)

Reviewer 4: ```suggestion
One important note is that the `-Yscala-release` flag is experimental and not yet available in any stable release of the compiler. It means the build tools, such as SBT or Mill, are not yet aware of its existence. Of course, you can declare usage of the flag in your build file, and the compiler will respect it. However, there may still be some hiccups, like generated artifacts depending on the incorrect version of the stdlib. We will do our best to help fix those minor issues before the stable release of 3.1.2.
```
?

Reviewer 2: this is improved. it's a bit hard for me to judge because I already know too much about the situation — I'd suggest running it by a few actual library maintainer or two who haven't been following this stuff closely

If you are interested in other changes brought by 3.1.2-RC1, there is, as always, a changelog on GitHub.  // TODO: link

## Contributors

// TODO: contributors list

## Library authors: Join our community build

Scala 3 has a set of widely-used community libraries that are built against every nightly Scala 3 snapshot.
Join our [community build](https://github.com/lampepfl/dotty/tree/master/community-build)

> to make sure that our regression suite includes your library.

Reviewer 2: I doubt this is sufficiently important to repeat in every release announcement.

Author: Sure, removed.

---
layout: blog-detail
post-type: blog
by: Paweł Marks, VirtusLab
title: Scala 3.1.1 and forward compatibility news
---

Hello from the Scala 3 team! We are back with a new release of the Scala compiler and exciting news about improvements in forward compatibility.

## Improvements is Scala 3.1.1

Scala 3.1.1 is the patch release focused on bug fixes. Although there are no new additions to the language, you can still spot some noticeable improvements:

- You can use `using` clauses to introduce subtyping relations between types in the function body.

  Now the code like this should compile. Any value of type `A` can be assigned to a value of type `B`:

  ```scala
    def example[A, B](a: A)(using A <:< B) =
      val b: B = a
      ???
  ```

  To know more see [PR #13662](https://github.com/lampepfl/dotty/pull/13662) and [related issue](https://github.com/lampepfl/dotty/issues/12955).

> - Mirrors can be correctly summoned for hierarchical sum types compiled by Scala 3.0.x. You can now safely create derived instances of type classes for the types defined in dependencies compiled with previous versions of the compiler. ([PR #14100](https://github.com/lampepfl/dotty/pull/14100)).

Reviewer 3: ```suggestion
- Mirrors can be correctly summoned for hierarchical sum types compiled by Scala 3.0.x. It is now safe to derive a type class for a type dependency that was compiled with a pre-3.1.0 compiler. ([PR #14100](https://github.com/lampepfl/dotty/pull/14100)).
```

We are sure we want to mitigate those difficulties. We already have implemented the first and most important part of the solution and have it ready for public testing. Beginning in Scala 3.1.2-RC1 compiler is able to generate output that can be consumed by older versions. To enable that, you need to specify a minimum version of the language using the experimental `-Yscala-release` flag.

We expect that in the future authors will use the newest available version of the compiler with `-Yscala-release` set to the lowest language version that support all features they need. This way they will be able to benefit from the all bugfixes without restricting their potential audience.

So, for example, compiling your library using Scala 3.1.2-RC2 with `-Yscala-release:3.0` flag, you make it possible for people still using Scala 3.0 to depend on your library. This has one limitation, you won't be able to use any symbol added to the standard library after 3.0. They are marked in the code and in the documentation by the `@since` annotation that was also added in this version.

> One important note is that the `-Yscala-release` flag is experimental and not yet available in any stable release of the compiler. It means the build tools, such as SBT or Mill, are not yet aware of its existence. Of course, you can declare usage of the flag in your build file, and the compiler will respect it. However, there may still be some hiccups, like generated artifacts depending on the incorrect version of the stdlib. We will do our best to help fix those minor issues before the stable release of 3.1.1.

Reviewer 5: > , like generated artifacts depending on the incorrect version of the stdlib

Could you specify exactly why this happens? MAybe `potential problems` section?

```suggestion
One important note is that the `-Yscala-release` flag is experimental and not yet available in any stable release of the compiler. It means the build tools, such as SBT or Mill, are not yet aware of its existence. Of course, you can declare usage of the flag in your build file, and the compiler will respect it. However, there may still be some hiccups, like generated artifacts depending on the incorrect version of the stdlib. We will do our best to help fix those minor issues before the stable release of 3.1.2.
```
?
